v1beta1 GKECluster

hasheddan opened the pull request:
<!--
Thank you for helping to improve Crossplane!

We strongly recommend you look through our contributor guide at https://git.io/fj2m9
if this is your first time opening a Crossplane pull request. You can find us in
https://slack.crossplane.io/messages/dev if you need any help contributing.
-->

### Description of your changes
<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

Fixes #
-->

Opening early to get feedback on implementation and open questions.

TODO:
- [x] Implement unit tests for v1beta1 GKECluster managed reconciler
- [x] Implement unit tests for v1beta1 GKE client pkg
- [x] Address cluster naming issues -- GCP prefers use of `projects/*/locations/*/clusters/*`
- [x] Add creation and immediate deletion of default NodePool with 0 pods (required to successfully provision a GKECluster
- [x] Refine GKECluster `AtProvider` status fields
- [x] Clean up field comments in GKECluster resource API
- ~~[ ] Clean up any duplicate code between `v1alpha3` and `v1beta1` (specifically with connection secret generation)~~
- ~~[ ] Add and test conversion functions for `v1alpha3` and `v1beta1 `GKECluster` versions (more info [here](https://book.kubebuilder.io/multiversion-tutorial/api-changes.html))~~

Open Questions:
- How to handle multiple updates needed for cluster? Since only one update can be performed at a time, the current strategy is to return the need for an update the first time an field that needs update is required, then check again after updating that one field (or object).
  - Answer: implemented as above for now
- How do we handle embedded structs where some fields are "output only" and some can be set in the spec? My immediate thought would be to just remove the output only fields from the struct and create an `Observed____` version of the struct that included the output fields.
  - Answer: Spec fields should be `___Spec` struct and Status fields in `___Status` per recommendation from @muvaf 
- How do we handle `Regional` vs `Zonal` clusters? The easiest way would be to add a `Location` field to the spec in addition to `Locations` even though it is not technically part of the `ForProvider` API spec. The `ProjectID` is also needed for all API calls so we will likely need to solve this with setting the `ExternalName` as `projects/{project_id_on_provider}/locations/{location_field_on_spec}/clusters/{cluster_name_in_spec}`
  - Answer: the solution described has been implemented.

### Checklist
<!--
Please run through the below readiness checklist. The first two items are
relevant to every Crossplane pull request.
-->
I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation], [examples], or [release notes].
- [ ] Updated the dependencies in [`app.yaml`] to include any new role permissions.

[documentation]: https://github.com/crossplaneio/crossplane/tree/master/docs
[examples]: https://github.com/crossplaneio/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplaneio/crossplane/tree/master/PendingReleaseNotes.md
[`app.yaml`]: https://github.com/crossplaneio/stack-gcp/blob/master/config/stack/manifests/app.yaml

hasheddan commented on apis/compute/v1beta1/gkecluster_types.go:
```diff
@@ -0,0 +1,1260 @@
+	// the cluster will be Regional, if a zone is given the cluster will be
+	// Zonal.
+	// +immutable
+	Location string `json:"location"`
```
`Location` has been added to struct despite it being listed as `Output Only` because it is needed in the API call parameters, not the `*container.Cluster` object.

hasheddan commented on pkg/controller/compute/v1beta1/cluster.go:
```diff
@@ -0,0 +1,287 @@
+	r := resource.NewManagedReconciler(mgr,
+		resource.ManagedKind(v1beta1.GKEClusterGroupVersionKind),
+		resource.WithExternalConnecter(&clusterConnector{kube: mgr.GetClient(), newServiceFn: container.NewService}),
+		resource.WithManagedInitializers())
```
Overriding default `NewManagedNameAsExternalName` initializer.

negz replied:
Do we need to do this?

I notice we're using the fully qualified name of the cluster (e.g. `projects/coolproj/locations/us-central1/clusters/mycluster`) rather than the unqualified name (e.g. `mycluster`). Are you perhaps thinking of the external name as meaning "the value that must be passed to the external client to identify the thing"? It's more intended to be "a setting the user can specify to control the name of their cluster as it appears in the GCP console".

My recommendation would be to use the unqualified name as the external name, use the regular external name initializer, and pass the result of some function with a signature like like `GetFullyQualifiedName(mg)` to the GKE client.

hasheddan replied:
I think I originally implemented this prior to adding the `Location` field, but you are definitely correct in that we can and should use the regular external name initializer here. Thanks for pointing that out!

hasheddan commented on pkg/controller/compute/v1beta1/cluster.go:
```diff
@@ -0,0 +1,287 @@
+	}
+
+	// If boostrap node pool is present then we want to delete it.
+	for _, pool := range existing.NodePools {
```
Check if 0 node "bootstrap" pool exists. If so we delete it. Alternatively, we could just try to delete it on every observation and ignore "not found" errors, but seems better to perform local operations rather than have additional network call in every reconciliation loop.

hasheddan commented on pkg/controller/compute/v1beta1/cluster.go:
```diff
@@ -0,0 +1,287 @@
+		Cluster: cluster,
+	}
+
+	parent := fmt.Sprintf(parentFormat, e.projectID, i.Spec.ForProvider.Location)
```
Using provided `Location` as parameter in `Create` call.

hasheddan commented on pkg/controller/compute/v1beta1/cluster.go:
```diff
@@ -0,0 +1,287 @@
+	}
+
+	clusterName := fmt.Sprintf(clusterNameFormat, e.projectID, i.Spec.ForProvider.Location, i.Spec.ForProvider.Name)
+	meta.SetExternalName(i, clusterName)
```
Set external name. On next `Observe` we should be able to successfully get the cluster with `meta.GetExternalName()`.

hasheddan commented on pkg/clients/container/cluster.go:
```diff
@@ -0,0 +1,996 @@
+		EnableTpu:             gcp.BoolValue(in.EnableTpu),
+		InitialClusterVersion: gcp.StringValue(in.InitialClusterVersion),
+		LabelFingerprint:      gcp.StringValue(in.LabelFingerprint),
+		Locations:             in.Locations,
```
Note: we do not set `Location` here because it is only to be used as a parameter in `Create()`, not in the actual `*container.Cluster` itself (output only).

hasheddan commented on apis/compute/v1beta1/gkecluster_types.go:
```diff
@@ -0,0 +1,1260 @@
+	// management of node pools is handled by the stack-gcp NodePool object.
+	// TODO(hasheddan): determine if we want to reflect node pools in the
+	// cluster status.
+	// NodePools []*NodePoolClusterStatus `json:"nodePools,omitempty"`
```
Do we want to include `NodePools` in the status? We should not be managing them here, but it might be nice to show which pools are going to be deleted if you delete a `GKECluster`. Note that though the GCP has `[]*NodePool` we must use `[]*NodePoolClusterStatus` because we have our own representation of a `NodePool` that includes object metadata.

negz commented on pkg/controller/compute/v1beta1/cluster.go:
```diff
@@ -0,0 +1,287 @@
+	"strings"
+	"time"
+
+	"k8s.io/client-go/tools/clientcmd"
```
Random floating import spotted.

negz commented on pkg/controller/compute/v1beta1/cluster.go:
```diff
@@ -0,0 +1,287 @@
+
+// Amounts of time we wait before requeuing a reconcile.
+const (
+	aLongWait         = 60 * time.Second
```
I don't think this is used.

negz commented on pkg/controller/compute/v1beta1/cluster.go:
```diff
@@ -0,0 +1,287 @@
+	requeueOnSucces = 2 * time.Minute
+
+	updateErrorMessageFormat         = "failed to update cluster object: %s"
+	erroredClusterErrorMessageFormat = "gke cluster is in %s state with message: %s"
```
We can probably write uppercase GKE here. Usually the rule is that error strings aren't capitalised, but acronyms get an exception.

negz commented on pkg/controller/compute/v1beta1/cluster.go:
```diff
@@ -0,0 +1,287 @@
+	cr.Status.AtProvider = gke.GenerateObservation(*existing)
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	gke.LateInitializeSpec(&cr.Spec.ForProvider, *existing)
+	if !reflect.DeepEqual(currentSpec, &cr.Spec.ForProvider) {
```
Should we use cmp.Equal here as we do in the client comparison functions?

hasheddan replied:
That sounds fine to me. I had noticed @muvaf used `DeepEqual` for redis and cloudsql, @muvaf was there any specific reason for that over `cmp.Equal`?

muvaf replied:
Here is why cmp package exists:
```
// This package is intended to be a more powerful and safer alternative to
// reflect.DeepEqual for comparing whether two values are semantically equal.
//
// The primary features of cmp are:
//
// • When the default behavior of equality does not suit the needs of the test,
// custom equality functions can override the equality operation.
// For example, an equality function may report floats as equal so long as they
// are within some tolerance of each other.
//
// • Types that have an Equal method may use that method to determine equality.
// This allows package authors to determine the equality operation for the types
// that they define.
//
// • If no custom equality functions are used and no Equal method is defined,
// equality is determined by recursively comparing the primitive kinds on both
// values, much like reflect.DeepEqual. Unlike reflect.DeepEqual, unexported
// fields are not compared by default; they result in panics unless suppressed
// by using an Ignore option (see cmpopts.IgnoreUnexported) or explicitly compared
// using the AllowUnexported option.
```

I remember thinking that I don't need any of those features, at least at that time, so I went with using the standard library for non-test code. But it should be fine to use it if you'd like. It's imported for the tests anyway, though I'm not 100% sure that if tests import a package it ends up compiled in the final binary.

negz replied:
I like that `cmp` ignores unexported fields, and allows custom equality override logic. Given `cmp` and `DeepEqual` do the same thing in most case, but `cmp` _can_ do more, my preference is to just default to using `cmp` for comparison.

negz commented on pkg/controller/compute/v1beta1/cluster.go:
```diff
@@ -0,0 +1,287 @@
+	// If boostrap node pool is present then we want to delete it.
+	for _, pool := range existing.NodePools {
+		if pool != nil {
+			if pool.Name == gke.BootstrapNodePoolName {
```
Does this if clause need to be nested inside the previous one?

I think you could safely write either:

```go
for _, pool := range existing.NodePools {
    // The second condition will never be assessed if the first is false.
    if pool != nil && pool.Name == gke.BootstrapNodePoolName {
        // Logic goes here...
    }
}
```

Or my personal preference in cases where the entire logic of a for block is inside an if is to use "continue early" to keep the happy path logic as un-nested as possible.

```go
for _, pool := range existing.NodePools {
    if pool == nil || pool.Name != gke.BootstrapNodePoolName {
        continue
    }
    // Logic goes here....
}
```

negz commented on pkg/controller/compute/v1beta1/cluster.go:
```diff
@@ -0,0 +1,287 @@
+			if pool.Name == gke.BootstrapNodePoolName {
+				boostrapNP := strings.Join([]string{meta.GetExternalName(mg), gke.BootstrapNodePoolName}, "/")
+				_, err = e.cluster.Projects.Locations.Clusters.NodePools.Delete(boostrapNP).Context(ctx).Do()
+				if err != nil {
```
Nit: this could be `if _, err := e.cluster...`

negz commented on pkg/controller/compute/v1beta1/cluster.go:
```diff
@@ -0,0 +1,287 @@
+		cr.Status.SetConditions(v1alpha1.Unavailable())
+	}
+
+	upToDate, _ := gke.IsUpToDate(&cr.Spec.ForProvider, *existing)
```
Style nitpick: this is a pretty detailed variable name for something that is returned from a method named `IsUpToDate` and used four lines later as the argument to a filed named `ResourceUpToDate`.

negz commented on pkg/controller/compute/v1beta1/cluster.go:
```diff
@@ -0,0 +1,287 @@
+		}
+	}
+
+	// If boostrap node pool is present then we want to delete it.
```
I think this would be a great place to use commentary to provide a little more context around what the bootstrap nodepool is and why we want to delete it. I suspect it would be pretty surprising to the uninitialised reader that observe deletes a nodepool.

...Speaking of which, would it make more sense for the Update method to own deleting the bootstrap nodepool?

hasheddan replied:
I like the idea of moving this to the `Update` method as that would mean we could handle the logic of identifying and deleting the node pool in `IsUpToDate()`. Thanks for the suggestion :)

negz commented on pkg/controller/compute/v1beta1/cluster.go:
```diff
@@ -0,0 +1,287 @@
+}
+
+func (e *clusterExternal) Create(ctx context.Context, mg resource.Managed) (resource.ExternalCreation, error) {
+	i, ok := mg.(*v1beta1.GKECluster)
```
I think @muvaf has been slowly renaming all these variables to `cr` for "custom resource". We might want to be consistent here.

(I personally used try to use a short version of the thing's name, which would be `c` in this case since it's a "GKE cluster" and not a "GKE instance", but I'm happy to settle on `cr`.)

negz commented on pkg/controller/compute/v1beta1/cluster.go:
```diff
@@ -0,0 +1,287 @@
+		return resource.ExternalUpdate{}, nil
+	}
+
+	_, err = updateFn(e.cluster, ctx, meta.GetExternalName(mg))
```
Is there a way we could avoid leaking knowledge of the GKE service, network context, etc into our clients package? I prefer to keep all that contained here in the controller code (or more specifically, prefer to have that knowledge and responsibility in one package rather than split across two).

Perhaps `IsUpToDate` could return the required `&container.UpdateClusterRequest{}` rather than returning a function that makes the actual update call?

hasheddan replied:
Unfortunately the reason this is happening is that some fields require specific updates methods, so `IsUpToDate()` does not always return an update func that is the same or even uses `&container.UpdateClusterRequest`

negz commented on pkg/controller/compute/v1beta1/cluster.go:
```diff
@@ -0,0 +1,287 @@
+	if gcp.IsErrorNotFound(err) {
+		return nil
+	}
+	return errors.Wrap(err, errDeleteCluster)
```
Nitpick: This could be a one liner like the below.

```go
_, err := // e.cluster...
return errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errDeleteCluster
```

hasheddan commented on apis/compute/v1beta1/gkecluster_types.go:
```diff
@@ -0,0 +1,1278 @@
+	// to create a cluster. It is not included in the actual cluster object
+	// itself, but is instead passed to the create call. If a region is given
+	// the cluster will be Regional, if a zone is given the cluster will be
+	// Zonal.
```
On my `TODO` list to clean up all of these :)

hasheddan commented on apis/compute/v1beta1/gkecluster_types.go:
```diff
@@ -0,0 +1,1278 @@
+// +kubebuilder:printcolumn:name="STATUS",type="string",JSONPath=".status.bindingPhase"
+// +kubebuilder:printcolumn:name="STATE",type="string",JSONPath=".status.state"
+// +kubebuilder:printcolumn:name="CLUSTER-NAME",type="string",JSONPath=".status.clusterName"
+// +kubebuilder:printcolumn:name="ENDPOINT",type="string",JSONPath=".status.endpoint"
```
Yep, these need to be cleaned up as well.

hasheddan commented on apis/compute/v1beta1/gkecluster_types.go:
```diff
@@ -0,0 +1,1346 @@
+	// This field is only applicable when `use_ip_aliases` is true.
+	// TODO(hasheddan): should this be removed?
+	// +optional
+	CreateSubnetwork *bool `json:"createSubnetwork,omitempty"`
```
We can allow subnetworks to be created for ip allocation, but would that be a violation of having subnetworks created outside of the `Subnetwork` resource? I would tend to think this is a special case that we could allow as part of the `GKECluster` configuration.

negz replied:
Yeah I feel okay with allowing this, especially given it doesn't have any config or knobs for managing subnets short of "make one". I believe the subnet is tied to the lifecycle of the cluster in this case, too.

hasheddan commented on apis/compute/v1beta1/gkecluster_types.go:
```diff
@@ -0,0 +1,1346 @@
+
+// NodePoolClusterStatus is a subset of information about NodePools associated
+// with a GKE cluster.
+type NodePoolClusterStatus struct {
```
This is a subset of the fields exposed for `NodePools` on a `GKECluster`. It could be argued that this is not "high fidelity", but since this is only present in status, a subset seems fine to me. It is the same fields that will be present in our `NodePoolObervation` struct + `Name` and `Version`

negz replied:
What's the thinking around only exposing a subset rather than all fields?

hasheddan replied:
Mostly just because we don't allow for configuration of fields here and so the primary objective seems to just be able to identify that a node pool exists and see its status. It also would add quite a bit of duplicate code in `NodePool` and `GKECluster`.

hasheddan commented on pkg/clients/container/cluster_test.go:
```diff
@@ -0,0 +1,1593 @@
+				upToDate: false,
+			},
+		},
+		"NoUpdateNotBootstrapNodePool": {
```
This and the subsequent test demonstrate how the update controller handles deletion of the default "bootstrap" node pool.

hasheddan commented on config/crd/compute.gcp.crossplane.io_gkeclusters.yaml:
```diff
@@ -38,342 +38,12 @@ spec:
-              description: State of this GKE cluster.
-              type: string
-          type: object
-      type: object
```
Not exactly sure how to get generation to reflect the schemas of both `v1alpha3` and `v1beta1` here :thinking:

hasheddan commented on apis/generate.go:
```diff
@@ -22,8 +22,14 @@ limitations under the License.
 // Remove existing CRDs
 //go:generate rm -rf ../config/crd
 
+// TODO(hasheddan): update following enabling version.
```
Note: generation is skipping over `v1beta1` for the time being

negz replied:
I'd suggest adding some more context to this TODO if we expect it to get merged - it's not quite clear what it means.

hasheddan commented on pkg/controller/compute/v1beta1/cluster.go:
```diff
@@ -0,0 +1,540 @@
+		out := &container.Cluster{}
+		gke.GenerateMasterAuth(in, out)
+		update := &container.SetMasterAuthRequest{
+			// TODO(hasheddan): need to set Action here?
```
Action is used to update `Username` / `Password` but we do not expose that to the user. We only expose whether we want a certificate to be generated, which I don't think can be undone so the change will be irrelevant, but you are able to attempt it, which is why it is being left in. Might be worth removing to avoid confusion though?

hasheddan replied:
I have confirmed that client `issueClientCertificate` is [unable to be updated](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/projects.locations.clusters/setMasterAuth) following creation. I am going to remove the `MasterAuth` update and mark `issueClientCertificate` as immutable given that it [can not be removed](https://cloud.google.com/kubernetes-engine/docs/how-to/hardening-your-cluster#restrict_authn_methods) after it is created.

muvaf commented on apis/compute/v1beta1/gkecluster_types.go:
```diff
@@ -0,0 +1,1686 @@
+	// +immutable
+	Location string `json:"location"`
+
+	// Locations: The list of Google Compute
```
I'm not sure I fully understand it but I guess GCP differentiates where the _cluster_ resides(`Location string`) and where the _nodes_ reside (`Locations []string`). Since we're going with `NodePool`s to define the nodes and [their locations](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1/projects.zones.clusters.nodePools/create), we probably don't need to have (`Locations []string`) here.

hasheddan replied:
`Location` is mostly used to distinguish if the `GKECluster` is `Zonal` or `Regional`. The `Location` field is actually used as a path parameter in the `Create` [call](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/projects.locations.clusters/create) and is only `[Ouput Only]` on the `*container.Cluster` object itself. `Locations []string` is part of the `*container.Cluster` and is relevant, especially if the the cluster is `Regional`. I am not certain but I believe you can set the zones in `Locations` and it will prevent `NodePool` creation attached to that cluster outside of those zones. It can also be used for selecting a subset of zones in a certain region to be used. I think it would be a violation of high fidelity to remove.

muvaf replied:
I don't really understand. For `locations` array, the doc says `The list of Google Compute Engine zones in which the cluster's nodes should be located. ` If we use inline or standalone `nodePools`, then each node pool has its own location field anyway. If this `locations` array is for inline deprecated `nodeConfig`, then why it's not deprecated as well?

> I believe you can set the zones in `Locations` and it will prevent NodePool creation attached to that cluster outside of those zones

I couldn't find this information in GCP docs or terraform docs. Anyway, I agree that we should keep it unless we get to know that it was there for deprecated inline nodeConfig.

muvaf commented on apis/compute/v1beta1/gkecluster_types.go:
```diff
@@ -0,0 +1,1686 @@
+
+// GKEClusterParameters define the desired state of a Google Kubernetes Engine
+// cluster.
+type GKEClusterParameters struct {
```
Not a strong opinion: I think the required fields could be at the top so that our doc generation tooling can put those fields in the top. Otherwise, it's a little bit hard to identify which fields are required.

muvaf commented on apis/compute/v1beta1/gkecluster_types.go:
```diff
@@ -0,0 +1,1686 @@
+	// +immutable
+	NetworkRef *NetworkURIReferencerForGKECluster `json:"networkRef,omitempty" resource:"attributereferencer"`
+
+	// NOTE(hasheddan): only intradnode visibility can be updated in
```
typo in `intradnode`

hasheddan commented on apis/compute/v1beta1/gkecluster_types.go:
```diff
@@ -0,0 +1,1686 @@
+
+// NOTE(hasheddan): The only auth configuration exposed is the ability to
+// generate client certificate. All auth output information is reflected in the
+// connection secret.
```
Good idea! `Username` and `Password` were recently removed to comply with decisions made on `CloudSQL` `v1beta1`.

hasheddan commented on apis/container/v1beta1/gkecluster_types.go:
```diff
@@ -0,0 +1,1685 @@
+// CloudRunConfig is configuration options for the Cloud Run feature.
+type CloudRunConfig struct {
+	// Disabled: Whether Cloud Run addon is enabled for this cluster.
+	Disabled *bool `json:"disabled"`
```
This is one where we `ForceSendFields`. It could technically not be a pointer because its parent is a pointer and it is required if its parent is, but I updated all fields that required force sending to pointer types to be consistent.

negz commented on pkg/controller/container/cluster.go:
```diff
@@ -0,0 +1,530 @@
+		resource.SetBindable(cr)
+	case v1beta1.ClusterStateProvisioning:
+		cr.Status.SetConditions(v1alpha1.Creating())
+	case v1beta1.ClusterStateUnspecified, v1beta1.ClusterStateDegraded, v1beta1.ClusterStateError, v1beta1.ClusterStateReconciling:
```
Is this cluster really unavailable in all of these states? What does reconciling mean?

hasheddan replied:
Reconciling is the status that is returned when some sort of update is being performed.

negz commented on pkg/controller/compute/v1beta1/cluster.go:
```diff
@@ -0,0 +1,540 @@
+	return errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errDeleteCluster)
+}
+
+// connectionSecret return secret object for cluster instance
```
In general I agree with @muvaf about putting more glue code in the client package, but I'm okay with the current implementation here.

If you think about the responsibilities each package has, frequently it's that the client package forms the data used to CRUD the external resource, while the controller package executes the actual update call. In some cases both responsibilities are in the client package, which presents its own client interface for the controller to call. Typically only one package knows anything about the actual cloud provider SDK calls, which is a good separation of responsibility between packages.

Initially in this PR the client package was forming the CRUD data, and the controller was executing the update calls, but due to the complex way in which GKE updates work the details of how GKE SDK was called were leaked into the client package. We solved this by moving that detail into the controller package - another (possibly better) option would have been to move _everything_ into the client package and have it present a CRUD API to the controller package, which would then not need to be aware of how the GKE SDK works. That was likely my fault, because my comment opened with "we should move everything into the controller package" before clarifying the actual intent.

So it _might_ be a little better to move everything into the client package, but I don't think it's worth the refactoring effort at this stage.

muvaf commented on pkg/clients/container/cluster.go:
```diff
@@ -0,0 +1,918 @@
+
+// GenerateNodePoolForCreate inserts the default node pool into
+// *container.Cluster so that it can be provisioned successfully.
+func GenerateNodePoolForCreate(in *container.Cluster) {
```
Yeah, `BootstrapNodePool` works perfectly, too. Just a nitpick, `Generate*` has an implication that something will be produced but here we're making an in-place change. So, maybe you can just say `AddNodePoolForCreate` or keep the name and make the function return a bootstrap node pool and assign it where you call the function.

muvaf commented on apis/compute/v1beta1/gkecluster_types.go:
```diff
@@ -0,0 +1,1686 @@
+/*
```
Got it, I just wanted to test it myself. Thanks for adding `managed.yaml`

hasheddan replied:
Cool! It actually will be claimable now that it is in the `container` group so I will be adding additional examples!

hasheddan commented:
I have pushed a few updates and done some fairly extensive manual testing:

Updates:
1. `name` field is removed in favor of using `external-name` (i.e. if statically provisioned the cluster is named the same as the CR, if dynamically and no `external-name` annotation set on claim then it is `<claim namespace>-<claim name>-<5 random chars>`).
1. `GKEClusterClass` `v1beta1` fixed to use `forProvider` in `specTemplate`.
1. `GKECluster` `v1beta1` claim controllers added to manager.
1. A few embedded struct single-child required fields have `omitempty` removed now.
1. Changed `GenerateNodePoolForCreate` to `AddNodePoolForCreate` per @muvaf request.
1. Added dynamic provisioning resource to `examples/` dir.
1. Updated printer columns to reflect fields in schema accurately.

I have successfully executed the following tests:
- Statically provisioned a *Regional* cluster
- Statically provisioned a *Zonal* cluster
- Statically provisioned a *Zonal* cluster with `networkRef` and `subnetworkRef` to `Network` and `Subnetwork` provisioned by Crossplane
- Claimed a statically provisioned cluster with a `KubernetesCluster` claim and confirmed that connection `Secret` is propagated to claim `Namespace` and resource and claim become `Bound`
- Dynamically provisioned a cluster using `KubernetesCluster and `v1beta1` `GKEClusterClass` and confirmed:
 - `GKECluster` `v1beta1` managed resource is created
 - GKE cluster is provisioned with name `<claim namespace>-<claim name>-<5 random chars>`
 - Connection `Secret` is propagated to claim namespace
 - Managed resource and claim become bound
- Updated `GKECluster` with various fields (including ones that require `ForceSendFields`) and observed updates taking place in and completing (updates take a long time but this is on the GCP side of things).
